File: db-4.8.30/examples_cxx/LockExample.cpp
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

#include <sys/types.h>

#include <errno.h>
#include <iostream>
#include <stdlib.h>
#include <string.h>

#include <db_cxx.h>

using std::cin;
using std::cout;
using std::cerr;

const char *progname = "LockExample";               // Program name.

//
// An example of a program using DBLock and related classes.
//
class LockExample : public DbEnv
{
public:
    void run();
    int error_code() { return (ecode); }

    LockExample(const char *home, u_int32_t maxlocks, int do_unlink);

private:
    static const char FileName[];
    int ecode;

    // no need for copy and assignment
    LockExample(const LockExample &);
    void operator = (const LockExample &);
};

static int usage();          // forward

int
main(int argc, char *argv[])
{
    const char *home;
    int do_unlink;
    u_int32_t maxlocks;
    int i;

    home = "TESTDIR";
    maxlocks = 0;
    do_unlink = 0;
    for (int argnum = 1; argnum < argc; ++argnum) {
        if (strcmp(argv[argnum], "-h") == 0) {
            if (++argnum >= argc)
                return (usage());
            home = argv[argnum];
        }
        else if (strcmp(argv[argnum], "-m") == 0) {
            if (++argnum >= argc)
                return (usage());
            if ((i = atoi(argv[argnum])) <= 0)
                return (usage());
            maxlocks = (u_int32_t)i;  /* XXX: possible overflow. */
        }
        else if (strcmp(argv[argnum], "-u") == 0) {
            do_unlink = 1;
        }
        else {
            return (usage());
        }
    }

    try {
        int ecode;

        if (do_unlink) {
            // Create an environment that immediately
            // removes all files.
            LockExample tmp(home, maxlocks, do_unlink);
            if ((ecode = tmp.error_code()) != 0)
                return (ecode);
        }

        LockExample app(home, maxlocks, do_unlink);
        if ((ecode = app.error_code()) != 0)
            return (ecode);
        app.run();
        app.close(0);
        return (EXIT_SUCCESS);
    }
    catch (DbException &dbe) {
        cerr << "LockExample: " << dbe.what() << "\n";
        return (EXIT_FAILURE);
    }
}

LockExample::LockExample(const char *home, u_int32_t maxlocks, int do_unlink)
:   DbEnv(0)
,   ecode(0)
{
    int ret;

    if (do_unlink) {
        if ((ret = remove(home, DB_FORCE)) != 0) {
            cerr << progname << ": DbEnv::remove: "
                 << strerror(errno) << "\n";
            ecode = EXIT_FAILURE;
        }
    }
    else {
        set_error_stream(&cerr);
        set_errpfx("LockExample");
        if (maxlocks != 0)
            set_lk_max_locks(maxlocks);
        open(home, DB_CREATE | DB_INIT_LOCK, 0);
    }
}

void LockExample::run()
{
    long held;
    size_t len;
    u_int32_t locker;
    int did_get, ret;
    DbLock *locks = 0;
    int lockcount = 0;
    int lockid = 0;
    char objbuf[1024];

    //
    // Accept lock requests.
    //
    lock_id(&locker);
    for (held = 0;;) {
        cout << "Operation get/release [get]> ";
        cout.flush();

        char opbuf[16];
        cin.getline(opbuf, sizeof(opbuf));
        if (cin.eof())
            break;
        if ((len = strlen(opbuf)) <= 1 || strcmp(opbuf, "get") == 0) {
            // Acquire a lock.
            cout << "input object (text string) to lock> ";
            cout.flush();
            cin.getline(objbuf, sizeof(objbuf));
            if (cin.eof())
                break;
            if ((len = strlen(objbuf)) <= 0)
                continue;

            char lockbuf[16];
            do {
                cout << "lock type read/write [read]> ";
                cout.flush();
                cin.getline(lockbuf, sizeof(lockbuf));
                if (cin.eof())
                    break;
                len = strlen(lockbuf);
            } while (len >= 1 &&
                 strcmp(lockbuf, "read") != 0 &&
                 strcmp(lockbuf, "write") != 0);

            db_lockmode_t lock_type;
            if (len <= 1 || strcmp(lockbuf, "read") == 0)
                lock_type = DB_LOCK_READ;
            else
                lock_type = DB_LOCK_WRITE;

            Dbt dbt(objbuf, (u_int32_t)strlen(objbuf));

            DbLock lock;
            ret = lock_get(locker, DB_LOCK_NOWAIT, &dbt,
                       lock_type, &lock);
            did_get = 1;
            lockid = lockcount++;
            if (locks == NULL) {
                locks = new DbLock[1];
            }
            else {
                DbLock *newlocks = new DbLock[lockcount];
                for (int lockno = 0;
                    lockno < lockid; lockno++) {
                    newlocks[lockno] = locks[lockno];
                }
                delete locks;
                locks = newlocks;
            }
            locks[lockid] = lock;
        } else {
            // Release a lock.
            do {
                cout << "input lock to release> ";
                cout.flush();
                cin.getline(objbuf, sizeof(objbuf));
                if (cin.eof())
                    break;
            } while ((len = strlen(objbuf)) <= 0);
            lockid = strtol(objbuf, NULL, 16);
            if (lockid < 0 || lockid >= lockcount) {
                cout << "Lock #" << lockid << " out of range\n";
                continue;
            }
            DbLock lock = locks[lockid];
            ret = lock_put(&lock);
            did_get = 0;
        }

        switch (ret) {
        case 0:
            cout << "Lock #" << lockid << " "
                 <<  (did_get ? "granted" : "released")
                 << "\n";
            held += did_get ? 1 : -1;
            break;
        case DB_LOCK_NOTGRANTED:
            cout << "Lock not granted\n";
            break;
        case DB_LOCK_DEADLOCK:
            cerr << "LockExample: lock_"
                 << (did_get ? "get" : "put")
                 << ": " << "returned DEADLOCK";
            break;
        default:
            cerr << "LockExample: lock_get: %s",
                strerror(errno);
        }
    }
    cout << "\n";
    cout << "Closing lock region " << held << " locks held\n";
    if (locks != 0)
        delete locks;
}

static int
usage()
{
    cerr << "usage: LockExample [-u] [-h home] [-m maxlocks]\n";
    return (EXIT_FAILURE);
}

File: db-4.8.30/examples_java/src/db/LockExample.java
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

package db;

import com.sleepycat.db.*;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.PrintStream;
import java.util.Vector;

//
// An example of a program using Lock and related classes.
//
class LockExample {
    private static final String progname = "LockExample";
    private static final File LOCK_HOME = new File("TESTDIR");
    Environment dbenv;

    public LockExample(File home, int maxlocks, boolean do_unlink)
        throws DatabaseException, FileNotFoundException {

        if (do_unlink) {
            Environment.remove(home, true, null);
        }

        EnvironmentConfig config = new EnvironmentConfig();
        config.setErrorStream(System.err);
        config.setErrorPrefix("LockExample");
        config.setMaxLocks(maxlocks);
        config.setAllowCreate(true);
        config.setInitializeLocking(true);
        dbenv = new Environment(home, config);
    }

    public void close() throws DatabaseException {
        dbenv.close();
    }

    // Prompts for a line, returning the default answer if a blank
    // line is entered.
    //
    public static String askForLine(InputStreamReader reader,
                                    PrintStream out, String prompt,
                                    String defaultAnswer) {
        String result;
        do {
            if (defaultAnswer != null)
                out.print(prompt + " [" + defaultAnswer + "] >");
            else
                out.print(prompt);
            out.flush();
            result = getLine(reader);
            if (result == null || result.length() == 0)
                result = defaultAnswer;
        } while (result == null);
        return result;
    }

    // Not terribly efficient, but does the job.
    // Works for reading a line from stdin or a file.
    // Returns null on EOF.  If EOF appears in the middle
    // of a line, returns that line, then null on next call.
    //
    public static String getLine(InputStreamReader reader) {
        StringBuffer b = new StringBuffer();
        int c;
        try {
            while ((c = reader.read()) != -1 && c != '\n')
                if (c != '\r')
                    b.append((char)c);
        } catch (IOException ioe) {
            c = -1;
        }

        if (c == -1 && b.length() == 0)
            return null;
        else
            return b.toString();
    }

    public void run() throws DatabaseException {
        long held;
        int len = 0, locker;
        int ret;
        boolean did_get = false;
        int lockid = 0;
        InputStreamReader in = new InputStreamReader(System.in);
        Vector locks = new Vector();

        //
        // Accept lock requests.
        //
        locker = dbenv.createLockerID();
        for (held = 0;;) {
            String opbuf = askForLine(in, System.out,
                                      "Operation get/release", "get");
            if (opbuf == null)
                break;

            try {
                if (opbuf.equals("get") || opbuf.equals("vget")) {
                    // Acquire a lock.
                    String objbuf = askForLine(in, System.out,
                           "input object (text string) to lock> ", null);
                    if (objbuf == null)
                        break;

                    String lockbuf;
                    do {
                        lockbuf = askForLine(in, System.out,
                                             "lock type read/write", "read");
                        if (lockbuf == null)
                            break;
                        len = lockbuf.length();
                    } while (len >= 1 &&
                             !lockbuf.equals("read") &&
                             !lockbuf.equals("write"));

                    LockRequestMode lock_type;
                    if (len <= 1 || lockbuf.equals("read"))
                        lock_type = LockRequestMode.READ;
                    else
                        lock_type = LockRequestMode.WRITE;

                    DatabaseEntry entry = new DatabaseEntry(objbuf.getBytes());

                    Lock lock;
                    if (opbuf.equals("get")) {
                        did_get = true;
                        lock = dbenv.getLock(locker, true, entry, lock_type);
                    } else {
                        LockRequest req = new LockRequest(
                            LockOperation.GET, lock_type, entry, null);
                        LockRequest[] reqs = { req };
                        dbenv.lockVector(locker, true, reqs);
                        lock = req.getLock();
                        System.out.println("Got lock: " + lock);
                    }
                    lockid = locks.size();
                    locks.addElement(lock);
                } else {
                    // Release a lock.
                    String objbuf;
                    objbuf = askForLine(in, System.out,
                                        "input lock to release> ", null);
                    if (objbuf == null)
                        break;

                    lockid = Integer.parseInt(objbuf, 16);
                    if (lockid < 0 || lockid >= locks.size()) {
                        System.out.println("Lock #" + lockid + " out of range");
                        continue;
                    }
                    did_get = false;
                    Lock lock = (Lock)locks.elementAt(lockid);
                    dbenv.putLock(lock);
                }
                System.out.println("Lock #" + lockid + " " +
                                   (did_get ? "granted" : "released"));
                held += did_get ? 1 : -1;
            } catch (LockNotGrantedException lnge) {
                System.err.println("Lock not granted");
            } catch (DeadlockException de) {
                System.err.println("LockExample: lock_" +
                                   (did_get ? "get" : "put") +
                                   ": returned DEADLOCK");
            } catch (DatabaseException dbe) {
                System.err.println("LockExample: lock_get: " + dbe.toString());
            }
        }
        System.out.println();
        System.out.println("Closing lock region " + String.valueOf(held) +
                           " locks held");
    }

    private static void usage() {
        System.err.println("usage: LockExample [-u] [-h home] [-m maxlocks]");
        System.exit(1);
    }

    public static void main(String[] argv) {
        File home = LOCK_HOME;
        boolean do_unlink = false;
        int maxlocks = 0;

        for (int i = 0; i < argv.length; ++i) {
            if (argv[i].equals("-h")) {
                if (++i >= argv.length)
                    usage();
                home = new File(argv[i]);
            } else if (argv[i].equals("-m")) {
                if (++i >= argv.length)
                    usage();

                try {
                    maxlocks = Integer.parseInt(argv[i]);
                } catch (NumberFormatException nfe) {
                    usage();
                }
            } else if (argv[i].equals("-u")) {
                do_unlink = true;
            } else {
                usage();
            }
        }

        try {
            LockExample app = new LockExample(home, maxlocks, do_unlink);
            app.run();
            app.close();
        } catch (DatabaseException dbe) {
            System.err.println(progname + ": " + dbe.toString());
        } catch (Throwable t) {
            System.err.println(progname + ": " + t.toString());
        }
        System.out.println("LockExample completed");
    }
}
